designer/server/src/main/scala/pl/touk/nussknacker/ui/db/timeseries/QuestDbFEStatisticsRepository.scala

```diff
@@ -0,0 +1,194 @@
+package pl.touk.nussknacker.ui.db.timeseries
+
+import cats.effect.{IO, Resource}
+import com.typesafe.scalalogging.LazyLogging
+import io.questdb.TelemetryConfiguration
+import io.questdb.cairo.security.AllowAllSecurityContext
+import io.questdb.cairo.wal.ApplyWal2TableJob
+import io.questdb.cairo.{CairoEngine, CairoException, DefaultCairoConfiguration}
+import io.questdb.griffin.{SqlException, SqlExecutionContextImpl}
+import io.questdb.log.LogFactory
+import io.questdb.std.Os
+import pl.touk.nussknacker.ui.db.timeseries.QuestDbFEStatisticsRepository.{createTableQuery, dropTablesQuery}
+
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Try, Using}
+
+// TODO list:
+// 1. Writing in a single thread and polish usage of the wal2table job.
+// 2. Configurable directory of db (for now it's tmp directory) and limiting db file space on disk.
+// 3. Is truncate ok on critical errors?
+// 4. Collecting statistics should be deactivable by configuration.
+// 5. API should have better types (missing domain layer for FE statistics names).
+// 6. Handling errors while creating the QuestDb.
+// 7. Write tests when db directory is deleted in runtime.
+// 8. Changing table definition and recreate
+private class QuestDbFEStatisticsRepository(private val cairoEngine: CairoEngine)
+  extends FEStatisticsRepository[Future]
+    with LazyLogging {
+  private val ctx = new SqlExecutionContextImpl(cairoEngine, 1).`with`(AllowAllSecurityContext.INSTANCE, null)
+  private val applyWal2TableJob = new ApplyWal2TableJob(cairoEngine, 1, 1)
+
+  private lazy val statsWalTableWriter = {
+    cairoEngine.getWalWriter(cairoEngine.getTableTokenIfExists("statistics"))
+  }
+
+  private lazy val recordCursorFactory = {
+    cairoEngine.select(QuestDbFEStatisticsRepository.selectQuery, ctx)
+  }
+
+  override def write(statistics: Map[String, Long])
+                    (implicit ec: ExecutionContext): Future[Unit] = withExceptionHandling(
+    () => {
+      statistics.foreach { entry =>
+        val row = statsWalTableWriter.newRow(Os.currentTimeMicros())
+        row.putStr(0, entry._1)
+        row.putLong(1, entry._2)
+        row.append()
+      }
+      statsWalTableWriter.commit()
```

Comment:
because of the blocking nature of the QuestDB (AFAIR), we should have a separate execution context to not make any unexpected, unintended influence on the designer. It seems that a fixed-size pool will be ok. But we have to be sure that if the pool exhausts (because of some problem with the QuestDB or our implementation), the errors won't be propagated in Futures returned by write/read methods. I see two potential solutions here:

1.  usage level recover - the code that uses the repository assumes that there can be some problem but the business level decision is to skip it and continue the main logic 
2. decorator level recover - we provide a decorator that is supposed to add the recovered logic around the read & write methods of `FEStatisticsRepository`. It would be better to create the decorator inside the class which uses the repository to explicitly show that ignoring the problems with the repository is a business decision.